Adding functionality for random data

ElizabethU opened the pull request:
# How to build and test:
In chef-load directory:
`hab package build .`
Unless there's a compile error, this builds a binary of the package you're working on. Find the hart file this command produced in the `results` folder, and move it to your a2 folder, with a command like the following
`cp results/elizabethuselton-chef-load-4.0.0-20171212193658-x86_64-linux.hart /Users/elizabethuselton/code/go/src/github.com/chef/a2/results/.`

Then, in an a2 studio:
`hab pkg install results/elizabethuselton-chef-load-4.0.0-20171212193658-x86_64-linux.hart  -b` but with whatever the hart file's name is.
Create a config file with `chef-load --sample-config > chef-load.conf`
After that, run `chef_load_nodes 10` to create 10 nodes.
`curl $ELASTICSEARCH_URL/node-state/_count` to see your number of nodes,
and `curl $ELASTICSEARCH_URL/node-state/_search` to see them.

# Open questions

1. Why is `organization_name` `demo` even with the randomized data? 
1. Why is `platform` `redhat` even with randomized data? 
Tentative answer: It seems it's pulling this from the `converge-success.json` and failing to override them, and `environment` is not specified in that file, meaning it doesn't get written over.
1. Has it always been this way? Possibly org and platform name have always been coming from `converge-success.json` even though they are supposed to be overwritten in the non-random data path by the same means that we are overwriting them in the random data path. https://github.com/chef/chef-load/blob/master/chef-client-run.go#L47
1. We know that chef_load is using `go-chef/chef`'s Node Struct. We are only seeing `Environment` get updated with our chef-load information, and we are wondering if this is because `Environment` is a top-level attribute in the Node Struct. None of the other fields are, though platform is allegedly in the ohai data which is getting saved as `AutomaticAttributes` but we can't figure out what exactly `AutomaticAttributes` is, where it comes from, or what it is composed of. 

Where We Left It:
1. Environment is successfully randomizing.
2. Organization and Platform are being randomized and integrated into the node data the same way they were in the non-random path, but they are not working. This is the current struggle.
3. Roles has code written for randomizing, but not hooked up, because we haven't found where in the node structure it should be. It's commented out so things compile.
4. No random seed has been added yet.

ElizabethU committed:
added copied random run generator function, with hard coded data

ElizabethU committed:
removed firstrun parameter

ElizabethU committed:
randomized platform

ElizabethU committed:
orgs and environments randomized

afiune reviewed (commented):
Looking good, I think we might need to change the Seed of the `rand` package so that it doesn't create the same pseudo-random numbers on every execution. Out of that, let's continue!!

afiune commented:
This card is progress of https://github.com/chef/a2/issues/241

rmoshier committed:
Sketch out roles

Commit:
Rename go files with dash to be underscore instead

Signed-off-by: Salim Afiune <afiune@chef.io>

Commit:
Updates for latest version of satori/go.uuid pkg

Signed-off-by: Salim Afiune <afiune@chef.io>

Commit:
Adds a single file with RandomData

This file contains all the raw data that chef-load will grab randomly,
if we need to add/remove raw data, do it here. Just one single place!

Signed-off-by: Salim Afiune <afiune@chef.io>

Commit:
Make ChefActions and ChefNodes use same RandomData

Signed-off-by: Salim Afiune <afiune@chef.io>

Commit:
Generate a random Start/End time

Signed-off-by: Salim Afiune <afiune@chef.io>

Commit:
Finishing adding functionality for random data

This commit does a bunch of things but in general it makes the random
data functionality to work!

Signed-off-by: Salim Afiune <afiune@chef.io>

afiune commented on generator.go:
```diff
@@ -87,3 +89,189 @@ func merge(cs ...<-chan error) <-chan error {
+
+	var (
+		startTime              = time.Now().UTC()
+		endTime                = time.Now().UTC() // TODO: (@afiune) should we mock the time the CCR took to run?
```
Do we wanna add this in next iterations?

afiune commented on generator.go:
```diff
@@ -87,3 +89,189 @@ func merge(cs ...<-chan error) <-chan error {
+
+	// uptime_seconds?
+	// If this is not set, the ingest pipeline explodes
+	// TODO: (@afiune) Fix it so it doesn't fail to ingest the node/ccr
```
I have a change that fixes this https://github.com/chef/a2/pull/575

afiune commented on random_data.go:
```diff
@@ -0,0 +1,178 @@
+
+package main
+
+// Random data to use by the generator.go file to create random nodes/ccr's
```
I really need help shaping this file, if we want something to be displayed in the UI, this is the file we have to modify the Raw Data. You can push changes or post ideas and I will commit them. (either way, thank you!!!)

lancewf commented on chef_actions.go:
```diff
@@ -208,8 +170,8 @@ func (ar *actionRequest) randomize() {
-	ar.ServiceHostname = randomServiceHostname()
-	ar.OrganizationName = randomOrganizationName()
+	ar.ServiceHostname = getRandom("source_fqdn")
+	ar.OrganizationName = getRandom("organization")
```
Thank you for changing this to use the Nodes values.

lancewf commented on generator.go:
```diff
@@ -87,3 +89,185 @@ func merge(cs ...<-chan error) <-chan error {
 	return out
 }
+
+func getRandom(kind string) string {
```
This is nice!

lancewf commented on generator.go:
```diff
@@ -87,3 +89,185 @@ func merge(cs ...<-chan error) <-chan error {
+func randomChefClientRun(chefClient chef.Client, nodeName string) {
+	var (
+		startTime              = time.Now().UTC()
+		endTime                = time.Now().UTC() // TODO: (@afiune) should we mock the time the CCR took to run?
```
Can the `endTime` be a few minutes after the `startTime`. The run-history component displays how long the run lasted. The `endTime` could also be a random amount of minutes from 1 to 10 after the `startTime`

lancewf reviewed (approved):
Looks good! I will get glad to see all the extra data.

afiune commented:
For screenshots of the UI with random data go to the issue [here](https://github.com/chef/a2/issues/241#issuecomment-355665406).

Commit:
Always return at least one recipe and attribute

Signed-off-by: Salim Afiune <afiune@chef.io>

Commit:
Generate random CCR status success/failure

Signed-off-by: Salim Afiune <afiune@chef.io>

kmacgugan reviewed (approved):
![tenor-5858483](https://user-images.githubusercontent.com/11166450/34629173-1c0a9b0a-f21c-11e7-841e-d5d1b55e95d5.gif)

afiune commented:
Habitat package built and uploaded:`✓ Uploaded afiune/chef-load/4.0.0/20180105212606`
